walkthroughs/week-4-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,64 @@
 
 package com.google.sps;
 
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Set;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> requestAttendees = request.getAttendees();
+    List<TimeRange> res = new LinkedList<>();
+    res.add(TimeRange.WHOLE_DAY);
+    for (Event e : events) {
+      Set<String> Attendees = e.getAttendees();
+      boolean flag = false;
+      for (String a : Attendees) {
+        if (requestAttendees.contains(a)) {
+          flag = true;
+          break;
+        }
+      }
+      if (!flag){
+        continue;
+      }
+      TimeRange badTime = e.getWhen();
+      ListIterator it = res.listIterator();
+      while (it.hasNext()) {
+        TimeRange goodTime = (TimeRange) it.next();
+        if (goodTime.start() >= badTime.start() + badTime.duration())
+          continue;
+        if (goodTime.start() < badTime.start()) {
+          if (badTime.start() >= goodTime.start() + goodTime.duration()) {
+            continue;
+          } else if (badTime.start() + badTime.duration() >= goodTime.start() + goodTime.duration()) {
+            it.set(TimeRange.fromStartDuration(goodTime.start(), badTime.start() - goodTime.start()));
+          } else {
+            it.set(TimeRange.fromStartDuration(goodTime.start(), badTime.start() - goodTime.start()));
+            it.add(TimeRange.fromStartDuration(badTime.start() + badTime.duration(),
+                goodTime.start() + goodTime.duration() - badTime.start() - badTime.duration()));
+          }
+        } else {
+          if (badTime.start() + badTime.duration() >= goodTime.start() + goodTime.duration()) {
+            it.remove();
+          } else {
+            it.set(TimeRange.fromStartDuration(badTime.start() + badTime.duration(),
+                goodTime.start() + goodTime.duration() - badTime.start() - badTime.duration()));
+          }
+        }
+      }
```

Comment:
I think you can use TimeRange.contains to detect these 4 cases:

1. bad time contains good time
2. good time contains bad time
3. bad time contains good time start
4. good time contains bad time start